File: src/lib/utils/parseJson.ts
/**
 * parseJson.ts — Shared LLM JSON parsing utility
 *
 * Handles all common LLM response formats:
 *   - Plain JSON:             { ... }
 *   - Markdown fenced:        ```json\n{ ... }\n```
 *   - Fenced without lang:    ```\n{ ... }\n```
 *   - JSON buried in prose:   "Here is the result:\n{ ... }\nDone."
 */

export interface ParseJsonResult<T = unknown> {
  data: T | null;
  raw: string;
  cleaned: string;
  error: string | null;
}

/**
 * Strip markdown code fences and extract the first JSON object or array
 * from an LLM response string.
 */
export function cleanLlmJson(raw: string): string {
  // 1. Remove opening fence  (```json, ```JSON, ```, etc.)
  let s = raw.replace(/^```[a-zA-Z]*\n?/, "").trim();

  // 2. Remove closing fence
  s = s.replace(/```\s*$/, "").trim();

  // 3. If there is still fenced content embedded (multi-fence responses)
  //    strip all remaining fences
  s = s.replace(/```[a-zA-Z]*\n?/g, "").replace(/```/g, "").trim();

  // 4. Extract first JSON object {...} or array [...]
  const objMatch = s.match(/\{[\s\S]*\}/);
  const arrMatch = s.match(/\[[\s\S]*\]/);

  if (objMatch && arrMatch) {
    // Return whichever appears first in the string
    return objMatch.index! <= arrMatch.index! ? objMatch[0] : arrMatch[0];
  }

  return objMatch?.[0] ?? arrMatch?.[0] ?? s;
}

/**
 * Safely parse an LLM response as JSON.
 *
 * @param raw       - Raw LLM response string
 * @param nodeName  - Node/file name for logging (e.g. "FinancialsNode")
 * @returns         ParseJsonResult with .data (null on failure) and .error
 */
export function safeParseLlmJson<T = unknown>(
  raw: string,
  nodeName: string
): ParseJsonResult<T> {
  const cleaned = cleanLlmJson(raw);

  console.debug(`[${nodeName}] Raw LLM response:`, raw.slice(0, 300));
  console.debug(`[${nodeName}] Cleaned JSON:`, cleaned.slice(0, 300));

  try {
    const data = JSON.parse(cleaned) as T;
    return { data, raw, cleaned, error: null };
  } catch (err) {
    const errorMsg = err instanceof Error ? err.message : String(err);
    console.error(`[${nodeName}] JSON parse error: ${errorMsg}`);
    console.error(`[${nodeName}] Problematic cleaned string:`, cleaned.slice(0, 500));
    return { data: null, raw, cleaned, error: errorMsg };
  }
}

/**
 * Extract the raw text content from a LangChain model response.
 * Works for both string responses and complex content arrays.
 */
export function extractTextContent(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content) && content.length > 0) {
    const first = content[0];
    if (first && typeof first === "object" && "text" in first) {
      return String((first as { text: unknown }).text);
    }
  }
  return "";
}

File: backend/utils/parse_json.py
"""
parse_json.py — LLM JSON parsing utility
Port of src/lib/utils/parseJson.ts

Handles all common LLM response formats:
  - Plain JSON:             { ... }
  - Markdown fenced:        ```json\\n{ ... }\\n```
  - Fenced without lang:    ```\\n{ ... }\\n```
  - JSON buried in prose:   "Here is the result:\\n{ ... }\\nDone."
"""

import re
import json
import logging
from typing import Any, Optional, Tuple

logger = logging.getLogger(__name__)


def clean_llm_json(raw: str) -> str:
    """Strip markdown code fences and extract the first JSON object or array."""
    # 1. Remove opening fence (```json, ```JSON, ```, etc.)
    s = re.sub(r"^```[a-zA-Z]*\n?", "", raw).strip()
    # 2. Remove closing fence
    s = re.sub(r"```\s*$", "", s).strip()
    # 3. Strip any remaining fences (multi-fence responses)
    s = re.sub(r"```[a-zA-Z]*\n?", "", s)
    s = re.sub(r"```", "", s).strip()

    # 4. Extract first JSON object {...} or array [...]
    obj_match = re.search(r"\{[\s\S]*\}", s)
    arr_match = re.search(r"\[[\s\S]*\]", s)

    if obj_match and arr_match:
        # Return whichever appears first
        return (
            obj_match.group()
            if obj_match.start() <= arr_match.start()
            else arr_match.group()
        )

    return obj_match.group() if obj_match else (arr_match.group() if arr_match else s)


def safe_parse_llm_json(raw: str, node_name: str) -> Tuple[Optional[Any], Optional[str]]:
    """
    Safely parse an LLM response as JSON.

    Args:
        raw:       Raw LLM response string.
        node_name: Node/file name for logging (e.g. "FinancialsNode").

    Returns:
        (data, error) where error is None on success.
    """
    cleaned = clean_llm_json(raw)

    logger.debug("[%s] Raw LLM response (first 300): %s", node_name, raw[:300])
    logger.debug("[%s] Cleaned JSON (first 300): %s", node_name, cleaned[:300])

    try:
        data = json.loads(cleaned)
        return data, None
    except json.JSONDecodeError as exc:
        error_msg = str(exc)
        logger.error("[%s] JSON parse error: %s", node_name, error_msg)
        logger.error("[%s] Problematic string: %s", node_name, cleaned[:500])
        return None, error_msg
